packages/components/src/components/SquiggleViewer/ViewerProvider.tsx

```diff
@@ -362,9 +361,54 @@ export function useViewerType() {
   return viewerType;
 }
 
+// List of callbacks that should do something outside of the viewer.
+// The common case is to scroll the editor to the necessary position in the playground.
+export type ExternalViewerActions = Partial<{
+  // TODO: this function is not imports-friendly yet.
+  // E.g. in stacktraces, the user might want to click on an error location and
+  // go to the source that's identified not just by an offset, but also its
+  // sourceId.
+  show: (location: SqLocation, focus: boolean) => void;
+  // When the viewer can focus on the location, will it be able to do that?
+  // This affects, for example, whether the links in stacktraces are clickable.
+  isFocusable: (location: SqLocation) => boolean;
+  // Should the viewer omit the source id because it's trivial?
+  // E.g. in playground stacktraces we want to show source id only for imports, but not for the currently edited file.
+  isDefaultSourceId: (sourceId: string) => boolean;
+}>;
+
+// Configure external actions for the common case of editor actions, e.g. for the playground or `<SquiggleEditor>` component.
+export function useExternalViewerActionsForEditor(
+  // Intentionally supports null and undefined - conditional hooks are not
+  // possible and this hook is often used in components that take an optional
+  // `editor` prop, or wire the editor components with a ref.
+  editor: CodeEditorHandle | null | undefined
+): ExternalViewerActions {
+  return useMemo(() => {
+    if (!editor) {
+      return {};
+    }
+    const actions: ExternalViewerActions = {
+      show: (location, focus) => {
+        // TODO - scroll to fit the full range on screen, and maybe select the entire range.
+        // TODO - check if sourceId matches the sourceId in editor.
+        editor.scrollTo(location.start.offset, focus);
+      },
+      isFocusable: (location) => {
+        // only the edited source id is focusable
+        return editor.getSourceId() === location.source;
+      },
+      isDefaultSourceId: (sourceId) => {
+        return editor.getSourceId() === sourceId;
+      },
+    };
+    return actions;
+  }, [editor]);
+}
+
 type Props = PropsWithChildren<{
   partialPlaygroundSettings: PartialPlaygroundSettings;
-  editor?: CodeEditorHandle;
+  externalViewerActions?: ExternalViewerActions;
```

Comment:
The prop name is intentionally verbose. At first I tried to shorten it to `actions` or `externalActions` when it was obvious from context, but it made things less readable and greppable.